File: solutions/python3/0090.py
class Solution:
    def subsetsWithDup(self, nums: List[int]) -> List[List[int]]:
        ret = []
        nums.sort()
        self.dfs(nums, len(nums), 0, [], ret)
        return ret

    def dfs(self, nums: List[int], target: int, s: int, curr: List[int], ret: List[List[int]]) -> None:
        if target < 0:
            return

        ret.append(curr)

        for i in range(s, len(nums)):
            if i > s and nums[i] == nums[i - 1]:
                continue
            self.dfs(nums, target - 1, i + 1, curr + [nums[i]], ret)

File: solutions/cpp/0090.cpp
class Solution {
public:
    vector<vector<int>> subsetsWithDup(vector<int>& nums) {
        vector<vector<int>> ans;
        vector<int> curr;
        sort(nums.begin(), nums.end());
        dfs(nums, nums.size(), 0, curr, ans);
        return ans;
    }

private:
    void dfs(vector<int>& nums, int target, int s, vector<int>& curr, vector<vector<int>>& ans) {
        if (target < 0) return;

        ans.push_back(curr);

        for (int i = s; i < nums.size(); i++) {
            if (i > s && nums[i] == nums[i - 1]) continue;
            curr.push_back(nums[i]);
            dfs(nums, target - 1, i + 1, curr, ans);
            curr.pop_back();
        }
    }
};